File: aqrar_ext/aqrar_ext/report/stock_ledger_report/stock_ledger_report.js
frappe.query_reports["Stock Ledger Report"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company") || frappe.defaults.get_default("Company"),
			reqd: 1,
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			reqd: 1,
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
		{
			fieldname: "warehouse",
			label: __("Warehouse"),
			fieldtype: "MultiSelectList",
			options: "Warehouse",
			get_data: function (txt) {
				return frappe.db.get_link_options("Warehouse", txt, {
					company: frappe.query_report.get_filter_value("company"),
				});
			},
		},
		{
			fieldname: "item_code",
			label: __("Item"),
			fieldtype: "MultiSelectList",
			options: "Item",
			get_data: async function (txt) {
				let { message: data } = await frappe.call({
					method: "erpnext.controllers.queries.item_query",
					args: { doctype: "Item", txt: txt, searchfield: "name", start: 0, page_len: 10, filters: {}, as_dict: 1 },
				});
				return (data || []).map(function (d) {
					return { value: d.name, description: Object.values(d).slice(1) };
				});
			},
		},
		{
			fieldname: "item_group",
			label: __("Item Group"),
			fieldtype: "Link",
			options: "Item Group",
		},
		{
			fieldname: "brand",
			label: __("Brand"),
			fieldtype: "Link",
			options: "Brand",
		},
		{
			fieldname: "batch_no",
			label: __("Batch No"),
			fieldtype: "Link",
			options: "Batch",
		},
		{
			fieldname: "voucher_no",
			label: __("Voucher #"),
			fieldtype: "Data",
		},
		{
			fieldname: "voucher_type",
			label: __("Transaction Type"),
			fieldtype: "Select",
			options: ["All", "Purchase Only", "Sale Only", "Transfer Only", "Stock Entry Only"],
			default: "All",
		},
		{
			fieldname: "project",
			label: __("Project"),
			fieldtype: "Link",
			options: "Project",
		},
		{
			fieldname: "include_uom",
			label: __("Include UOM"),
			fieldtype: "Link",
			options: "UOM",
		},
		{
			fieldname: "valuation_field_type",
			label: __("Valuation Field Type"),
			fieldtype: "Select",
			options: "Currency\nFloat",
			default: "Currency",
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		if (column.fieldname === "qty_in" && flt(value) > 0) {
			return "<span style='color:var(--green-600);font-weight:600'>" + value + "</span>";
		}
		if (column.fieldname === "qty_out" && flt(value) > 0) {
			return "<span style='color:var(--red-600);font-weight:600'>" + value + "</span>";
		}
		if (column.fieldname === "qty_after_transaction") {
			var color = flt(value) >= 0 ? "var(--blue-600)" : "var(--red-600)";
			return "<span style='color:" + color + ";font-weight:600'>" + value + "</span>";
		}
		return default_formatter(value, row, column, data);
	},

	onload: function (report) {
		report.page.add_inner_button(__("View Stock Balance"), function () {
			frappe.set_route("query-report", "Stock Balance", report.get_values());
		});
		_render_special_analysis_links(report);
	},
};

function _render_special_analysis_links(report) {
	var custom_reports = (window.aqrar_ext || {}).stock_ledger_special_reports || [];
	if (!custom_reports.length) return;

	var $wrapper = $(report.wrapper).find(".page-form");
	if ($wrapper.find(".aqrar-special-analysis").length) return;

	var $links = $(
		'<div class="aqrar-special-analysis" style="padding:8px 0 4px;">' +
			'<span style="font-weight:600;margin-right:8px;">' + __("Special Analysis:") + "</span>" +
		"</div>"
	);

	custom_reports.forEach(function (r) {
		$links.append(
			$('<a class="btn btn-xs btn-default" style="margin-right:6px;">')
				.text(__(r.label))
				.on("click", function () {
					frappe.set_route("query-report", r.report_name);
				})
		);
	});

	$wrapper.append($links);
}

File: aqrar_ext/aqrar_ext/report/stock_ledger_report/stock_ledger_report.py
import re

import frappe
from frappe import _
from frappe.utils import flt


def execute(filters=None):
	filters = filters or {}
	columns = get_columns(filters)
	data    = get_data(filters)
	return columns, data


def get_columns(filters=None):
	filters       = filters or {}
	val_fieldtype = filters.get("valuation_field_type") or "Currency"

	cols = [
		{"label": _("Date"),           "fieldname": "date",                  "fieldtype": "Data",         "width": 160},
		{"label": _("Item"),           "fieldname": "item_code",             "fieldtype": "Link",          "options": "Item",        "width": 130},
		{"label": _("Item Name"),      "fieldname": "item_name",             "fieldtype": "Data",          "width": 150},
		{"label": _("Stock UOM"),      "fieldname": "stock_uom",             "fieldtype": "Link",          "options": "UOM",         "width": 90},
		{"label": _("In Qty"),         "fieldname": "qty_in",                "fieldtype": "Float",         "width": 80},
		{"label": _("Out Qty"),        "fieldname": "qty_out",               "fieldtype": "Float",         "width": 80},
		{"label": _("Balance Qty"),    "fieldname": "qty_after_transaction", "fieldtype": "Float",         "width": 100},
	]

	if filters.get("include_uom"):
		cols += [
			{"label": _("UOM"),         "fieldname": "alt_uom",              "fieldtype": "Link",          "options": "UOM",         "width": 80},
			{"label": _("UOM Qty"),     "fieldname": "uom_qty",              "fieldtype": "Float",         "width": 90},
		]

	cols += [
		{"label": _("Warehouse"),      "fieldname": "warehouse",             "fieldtype": "Link",          "options": "Warehouse",   "width": 150},
		{"label": _("Voucher Type"),   "fieldname": "voucher_type",          "fieldtype": "Data",          "width": 120},
		{"label": _("Voucher #"),      "fieldname": "voucher_no",            "fieldtype": "Dynamic Link",  "options": "voucher_type","width": 160},
		{"label": _("Batch No"),       "fieldname": "batch_no",              "fieldtype": "Link",          "options": "Batch",       "width": 110},
		{"label": _("Project"),        "fieldname": "project",               "fieldtype": "Link",          "options": "Project",     "width": 110},
		{"label": _("Item Group"),     "fieldname": "item_group",            "fieldtype": "Link",          "options": "Item Group",  "width": 110},
		{"label": _("Brand"),          "fieldname": "brand",                 "fieldtype": "Data",          "width": 100},
		{"label": _("Description"),    "fieldname": "description",           "fieldtype": "Data",          "width": 150},
		{"label": _("Valuation Rate"), "fieldname": "valuation_rate",        "fieldtype": val_fieldtype,   "width": 120},
		{"label": _("Stock Value"),    "fieldname": "stock_value",           "fieldtype": val_fieldtype,   "width": 120},
	]
	return cols


def get_data(filters):
	conditions, values = _build_conditions(filters)
	where = " AND ".join(conditions)

	include_uom = filters.get("include_uom")

	uom_select = ""
	uom_join   = ""
	if include_uom:
		uom_select = ", uom_conv.uom AS alt_uom, uom_conv.conversion_factor AS uom_conv_factor"
		uom_join   = """
			LEFT JOIN `tabUOM Conversion Detail` uom_conv
				ON uom_conv.parent = sle.item_code
				AND uom_conv.uom = %(include_uom)s
		"""
		values["include_uom"] = include_uom

	rows = frappe.db.sql(f"""
		SELECT
			sle.posting_date,
			sle.posting_time,
			sle.item_code,
			i.item_name,
			i.stock_uom,
			i.item_group,
			i.brand,
			i.description,
			sle.warehouse,
			sle.voucher_type,
			sle.voucher_no,
			sle.batch_no,
			sle.project,
			sle.actual_qty,
			sle.qty_after_transaction,
			sle.valuation_rate,
			sle.stock_value
			{uom_select}
		FROM `tabStock Ledger Entry` sle
		LEFT JOIN `tabItem` i ON i.name = sle.item_code
		{uom_join}
		WHERE {where}
		ORDER BY sle.item_code, sle.posting_date, sle.posting_time, sle.creation
	""", values, as_dict=True)

	data = []
	for row in rows:
		qty = flt(row["actual_qty"])
		row["qty_in"]  = qty      if qty > 0 else 0
		row["qty_out"] = abs(qty) if qty < 0 else 0
		row["date"]    = str(row["posting_date"]) + " " + str(row["posting_time"] or "")[:8]

		if include_uom:
			conv = flt(row.get("uom_conv_factor") or 0)
			row["uom_qty"] = flt(row["qty_after_transaction"]) / conv if conv else 0
			row["alt_uom"] = include_uom if conv else None

		data.append(row)

	return data


def _build_conditions(filters):
	conditions = ["sle.is_cancelled = 0"]
	values     = {}

	if filters.get("company"):
		conditions.append("sle.company = %(company)s")
		values["company"] = filters["company"]

	if filters.get("from_date"):
		conditions.append("sle.posting_date >= %(from_date)s")
		values["from_date"] = filters["from_date"]

	if filters.get("to_date"):
		conditions.append("sle.posting_date <= %(to_date)s")
		values["to_date"] = filters["to_date"]

	# item_code: single value or MultiSelectList (comma-separated)
	item_codes = _parse_multiselect(filters.get("item_code"))
	if item_codes:
		if len(item_codes) == 1:
			conditions.append("sle.item_code = %(item_code)s")
			values["item_code"] = item_codes[0]
		else:
			conditions.append("sle.item_code IN %(item_codes)s")
			values["item_codes"] = tuple(item_codes)

	# warehouse: single value or MultiSelectList
	warehouses = _parse_multiselect(filters.get("warehouse"))
	if warehouses:
		if len(warehouses) == 1:
			conditions.append("sle.warehouse = %(warehouse)s")
			values["warehouse"] = warehouses[0]
		else:
			conditions.append("sle.warehouse IN %(warehouses)s")
			values["warehouses"] = tuple(warehouses)

	if filters.get("item_group"):
		conditions.append("i.item_group = %(item_group)s")
		values["item_group"] = filters["item_group"]

	if filters.get("brand"):
		conditions.append("i.brand = %(brand)s")
		values["brand"] = filters["brand"]

	if filters.get("batch_no"):
		conditions.append("sle.batch_no = %(batch_no)s")
		values["batch_no"] = filters["batch_no"]

	if filters.get("voucher_no"):
		conditions.append("sle.voucher_no = %(voucher_no)s")
		values["voucher_no"] = filters["voucher_no"]

	vt = filters.get("voucher_type") or "All"
	if vt and vt != "All":
		if vt == "Purchase Only":
			conditions.append("sle.voucher_type IN ('Purchase Invoice', 'Purchase Receipt')")
		elif vt == "Sale Only":
			conditions.append("sle.voucher_type IN ('Sales Invoice', 'Delivery Note')")
		elif vt == "Transfer Only":
			conditions.append("sle.voucher_type = 'Stock Entry' AND sle.voucher_no IN (SELECT name FROM `tabStock Entry` WHERE purpose IN ('Material Transfer', 'Material Transfer for Manufacture'))")
		elif vt == "Stock Entry Only":
			conditions.append("sle.voucher_type = 'Stock Entry'")

	if filters.get("project"):
		conditions.append("sle.project = %(project)s")
		values["project"] = filters["project"]

	return conditions, values


def _parse_multiselect(value):
	"""Normalise a filter value to a list.

	MultiSelectList sends a list; a Link/Data filter sends a plain string; and
	hand-typed values arrive newline- or comma-separated. The previous version
	split on newlines only, so "A,B" was treated as one item code.
	"""
	if not value:
		return []
	if isinstance(value, (list, tuple, set)):
		return [str(v).strip() for v in value if v and str(v).strip()]

	parts = re.split(r"[\n,]", str(value))
	return [p.strip() for p in parts if p.strip()]
